parser/src/parser/grammar.lalrpop

```diff
@@ -119,7 +158,11 @@ BoundaryAtom: BoundaryExpr = {
     <lexpr: BoundaryAtom> "^" <num: Num_u64> => BoundaryExpr::Exp(Box::new(lexpr), num),
     "$rand" "[" <n: Num_u64> "]" => BoundaryExpr::Rand(n as usize),
     <n: Num_u64> => BoundaryExpr::Const(n),
-    <ident: Identifier> "[" <index: Num_u64> "]" => BoundaryExpr::PubInput(ident, index as usize)
+    <ident: Identifier> => BoundaryExpr::Elem(ident),
+    <ident: Identifier> "[" <idx: Num_u64> "]" => 
+        BoundaryExpr::VecElem(VectorAccess::new(ident, idx)),
+    <ident: Identifier> "[" <row_idx: Num_u64> "]" "[" <col_idx: Num_u64> "]" =>
+        BoundaryExpr::MatrixElem(MatrixAccess::new(ident, row_idx, col_idx))
```

Comment:
Since most of this is reused for the transition expressions, I would pull most of it out of here so it's all in one place. Here's how I would approach this:

First, create 2 new atoms for indexing and size. These will be useful for column groupings as well. We can immediately each time we index to use `<idx: Index>` instead (e.g. indexing into `rand`). We can update the declaration of public inputs to use `Size`.
```
Size: u64 = {
    "[" <size: Num_u64> "]" => size
}

Index: usize = {
    "[" <idx: Num_u64> "]" => idx as usize
}
```

Next, I'd create rules for accessing vectors and matrices.
```
VectorAccess: VectorAccess = {
    <ident: Identifier> <idx: Index> => VectorAccess::new(ident, idx)
}

MatrixAccess: MatrixAccess = {
    <ident: Identifier> <row: Index> <col: Index> => MatrixAccess::new(ident, row, col)
}
```

Finally, I'd replace the boundary and transition constraint parsing rules to use `VectorAccess` and `MatrixAccess`. This will also make those rules easier to read.

```
BoundaryAtom: BoundaryExpr = {
    "(" <BoundaryExpr> ")",
    <lexpr: BoundaryAtom> "^" <num: Num_u64> => BoundaryExpr::Exp(Box::new(lexpr), num),
    "$rand" <n: Index> => BoundaryExpr::Rand(n as usize),
    <n: Num_u64> => BoundaryExpr::Const(n),
    <ident: Identifier> => BoundaryExpr::Elem(ident),
    <vector_access: VectorAccess> => BoundaryExpr::VecElem(vector_access),
    <matrix_access: MatrixAccess> => BoundaryExpr::MatrixElem(matrix_access)
}
```

Reply:
Thank you. Fixed.